theano/gpuarray/opt.py

```diff
@@ -315,7 +328,12 @@ def apply(self, fgraph):
 
             # Move only if any of the inputs are on the GPU.
             move_to_GPU = False
+            if any([isinstance(i, GpuArrayVariable) or
+                   isinstance(i, GpuArraySharedVariable)
+                   for i in [mapping[v] for v in node.inputs] +
+                   node.outputs]):
 
+                move_to_GPU = True
```

Comment:
Why did you add it back? I think it is not needed. I think the loop at line 338 in this commit does it for the inputs and for the output, I think it is not useful as we do one pass from the input to the output. The output won't be on the GPU unless the op already work on the GPU.